src/sagemaker_containers/_intermediate_output.py

```diff
@@ -0,0 +1,177 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from __future__ import absolute_import
+
+import concurrent.futures as futures
+import multiprocessing
+import os
+import shutil
+import time
+
+import boto3
+import boto3.s3.transfer as s3transfer
+import inotify_simple
+from six.moves.urllib.parse import urlparse
+
+from sagemaker_containers import _env, _logging
+
+logger = _logging.get_logger()
+
+intermediate_path = _env.output_intermediate_dir  # type: str
+failure_file_path = os.path.join(_env.output_dir, 'failure')  # type: str
+success_file_path = os.path.join(_env.output_dir, 'success')  # type: str
+tmp_dir_path = os.path.join(intermediate_path, '.tmp.sagemaker_s3_sync')  # type: str
+
+
+def _timestamp():
+    """Return a timestamp with microsecond precision."""
+    moment = time.time()
+    moment_us = repr(moment).split('.')[1]
+    return time.strftime("%Y-%m-%d-%H-%M-%S-{}".format(moment_us), time.gmtime(moment))
+
+
+def _upload_to_s3(s3_uploader, relative_path, file_path, filename):
+    try:
+        key = os.path.join(s3_uploader['key_prefix'], relative_path, filename)
+        s3_uploader['transfer'].upload_file(file_path, s3_uploader['bucket'], key)
+
+        # delete the original file
+        os.remove(file_path)
+    except FileNotFoundError:
+        # Broken link or deleted
+        pass
+    except Exception:
+        logger.exception('Failed to upload file to s3.')
+
+        # delete the original file
+        os.remove(file_path)
```

Comment:
this is in try and except branch. move to finally (with exists check to avoid another FNF